### ISSUE TITLE
IME not working on mobile

### ISSUE THREAD

**Issue by jackycute** [codemirror/CodeMirror#3137]
I upgraded to version 5.0 several days ago.
But I found that I cannot type chinese on the mobile.
It will show the raw Bopomofo like this "ㄋ一".
Which should only stay in IME or composition event.

BTW, if I force to use "textarea" mode in inputStyle config.
It can type chinese correctly but not the auto complete from built-in keyboard.
If you do so, It will lost focus and we need to click other lines again to type, or nothing will show up.

Thanks for any support, CodeMirror is awesome!

**Comment by marijnh** [comment]
That's interesting. Do you see the same problem on desktop CodeMirror when you explicitly enable the `"contenteditable"` input style?

One possibility is that `input.forceCompositionEnd` is being called when it shouldn't. Could you try to verify that with a `console.log`? Another possibility is that the IME software is getting confused by some aspect of CodeMirror's DOM behavior, and is itself prematurely ending the composition because of that.

I unfortunately know very little about both Chinese writing and Chinese IME. For me to debug this you'd have to walk me through the procedure of enabling it on the mobile platform you are using, and explain what you are doing and what the expected outcome is. (If the problem also manifests itself on the desktop, that would help. I have a Japanese IME installed, but a quick test didn't show any problem using that with contenteditable mode.)

**Comment by jackycute** [comment]
No, desktop version using "contenteditable" input type won't cause this, but using this mode in Chrome is extremely slow, I don't know why.

The good new is, I found I can debug on mobile via Safari, which my Mac connected my iPhone.

Let's talk about "textarea" mode in mobile first.
The real problem is, when we input in composition, the backspace action didn't trigger any composition event, and instead, CodeMirror found you input the backspace and clear its textarea, which cause the built-in autocomplete failed and lost focus to where I didn't know.
Then I need to click finish on my keyboard and click to another line to type again.

I haven't check about what break "contenteditable" in mobile version, maybe as you said.
I will report it later when I check it, thanks for your help! Let's make CodeMirror better!

**Comment by lackneets** [comment]
I have the same problem. 
Here I found a hack solution: not pretty well but works
https://github.com/zhusee2/codemirror-composition-mod

**Comment by jackycute** [comment]
Tks! @lackneets

**Comment by episodeyang** [comment]
Hey guys:)

Has there been any update on this? If Jacky's solution is good, maybe we can incorporate it into the core code base?

**Comment by marijnh** [comment]
I suspect the issue is that Safari/Mobile Safari are very easily confused when a JavaScript program messed with the editable DOM, and will either get confused (as in the auto-capitalizing of the keyboard) or fail to treat the updated content correctly (similar problems exist for spell check and double-space-for-period). In ProseMirror (a rich text editor) I've managed to work around this by letting the native behavior through and only updating the DOM when absolutely necessary. Eventually, I want to move CodeMirror's contenteditable mode to this approach as well, and I suspect it will help for this issue, but that's going to be a bunch of work it might take a while before I get to it.

As for the mod at https://github.com/zhusee2/codemirror-composition-mod , that looks like it'll cause more problems than it solves (it resets the DOM during composition, which will often force the composition to end, and it blindly trusts the data in the composition events, which is inaccurate on many platforms).

**Comment by episodeyang** [comment]
Hey @Marijnh! How is it going : )

Yeah, that sound reasonable. I will wait for these updates:)

**Comment by zxl777** [comment]
+1  I am waiting for these updates too ;)

**Comment by EveLuty** [comment]
is there anything new for that ? I found that problem also in my project.. if there are some update, it will be great. :+1:

**Comment by barretlee** [comment]
How is it going? seems still not working.

**Comment by barretlee** [comment]
`textarea` input style has lots of other problem, such as cannot select words.

I have an idea, we do not hidden `textarea`, rather put it under the rendered div, set opacity to 0, it may resolve the selection problem .

**Comment by adrianheine** [comment]
> Eventually, I want to move CodeMirror's contenteditable mode to this approach as well, and I suspect it will help for this issue, but that's going to be a bunch of work it might take a while before I get to it.

Just for clarity, this is something @marijnh would like to see happening, but currently neither has time nor funding to do so afaik.

**Comment by gfx** [comment]
The same in iOS Safari with Japanese IME.

**Comment by gfx** [comment]
With `inputStyle: "textarea"`, IME works well, but the native cursor for textarea also displays in addition to the CodeMirror cursor.

<img width="363" alt="screen shot 2017-07-27 at 12 03 14" src="https://user-images.githubusercontent.com/101800/28652599-a2dea96a-72c3-11e7-953f-158925e9c5d2.png">

The black one is CodeMirror's cursor while the blue one is the native textarea's.

**Comment by zcyzcy88** [comment]
Orion editor works on touch screen well (because Orion editor is directly use `contenteditable`, not simulate everything)

**Comment by inductor** [comment]
What's up with this issue? :s

**Comment by mhlx** [comment]
http://lab.lepture.com/editor/ 
this link solved the problem,but it is 3.5

**Comment by mhlx** [comment]
thank you @zakahiro

**Comment by adrianheine** [comment]
This is a issue that's difficult if not impossible to solve with the fundamental approach currently taken by CodeMirror.

We are working on a rewrite (CodeMirror 6) that will address this issue, and we are currently raising money for this work: See [the announcement](https://codemirror.net/6/) for more information about the rewrite and a demo.

Note that CodeMirror 6 is by no means stable or usable in production, yet. It's highly unlikely that we pick up this issue for CodeMirror 5, though.

**Comment by mhlx** [comment]
https://www.qyh.me/static/markdown-editor/index.html

**Comment by mhlx** [comment]
After I upgrade my iphone to 13,this bug is fixed!!!
